lib/index.js

```diff
@@ -72,388 +35,341 @@ const defaultRules = [
+      // .npmignore means no .gitignore
+      this.ignoreRules['.gitignore'] = null
+    }
 
-      // fall back to filesystem when seeded manifest is invalid
-      if (!pkg || typeof pkg !== 'object') {
-        return this.readPackageJson(entries)
-      }
+    return super.filterEntries()
+  }
 
-      // feels wonky, but this ensures package bin is _always_
-      // normalized, as well as guarding against invalid JSON
-      return this.getPackageFiles(entries, JSON.stringify(pkg))
+  // overridden method: we never want to include anything that isn't a file or directory
+  onstat (opts, callback) {
+    if (!opts.st.isFile() && !opts.st.isDirectory()) {
+      return callback()
     }
 
-    this.readPackageJson(entries)
+    return super.onstat(opts, callback)
   }
 
-  onReadPackageJson (entries, er, pkg) {
-    if (er) {
-      this.emit('error', er)
-    } else {
-      this.getPackageFiles(entries, pkg)
+  // overridden method: we want to refuse to pack files that are invalid, node-tar protects us from
+  // a lot of them but not all
+  stat (opts, callback) {
+    if (nameIsBadForWindows(opts.entry)) {
+      return callback()
     }
+
+    return super.stat(opts, callback)
   }
 
-  mustHaveFilesFromPackage (pkg) {
-    const files = []
-    if (pkg.browser) {
-      files.push('/' + pkg.browser)
-    }
-    if (pkg.main) {
-      files.push('/' + pkg.main)
+  // overridden method: we need to load the arborist tree before we actually start running
+  start () {
+    if (this.isPackage && !this.tree) {
+      const arborist = new Arborist({ path: this.path })
+      arborist.loadActual().then((tree) => {
+        this.tree = tree
+        super.start()
+      })
+      return this
     }
-    if (pkg.bin) {
-      // always an object because normalized already
-      for (const key in pkg.bin) {
-        files.push('/' + pkg.bin[key])
-      }
-    }
-    files.push(
-      '/package.json',
-      '/npm-shrinkwrap.json',
-      '!/package-lock.json',
-      packageMustHaves
-    )
-    return files
+
+    return super.start()
   }
 
-  getPackageFiles (entries, pkg) {
-    try {
-      // XXX this could be changed to use read-package-json-fast
-      // which handles the normalizing of bins for us, and simplifies
-      // the test for bundleDependencies and bundledDependencies later.
-      // HOWEVER if we do this, we need to be sure that we're careful
-      // about what we write back out since rpj-fast removes some fields
-      // that the user likely wants to keep. it also would add a second
-      // file read that we would want to optimize away.
-      pkg = normalizePackageBin(JSON.parse(pkg.toString()))
-    } catch (er) {
-      // not actually a valid package.json
-      return super.onReaddir(entries)
+  // overridden method: we don't want to respect package.json for children walkers, so we remove
+  // it from the ignoreFiles passed to children
+  walkerOpt (entry, opts) {
```

Comment:
It looks to me like you're entirely replacing ignoreFiles rather than removing anything specific?

Reply:
seemingly in contradiction to the comment

Reply:
yeah the comment is probably a bit unclear, i'll fix it. what's really happening is the `walkerOpt` method gets run to build a set of options when we step into a child directory. we don't want to respect any package.json files beyond the root and the default `walkerOpt` copies the root's `ignoreFiles`. we don't want that, so we set the list of `ignoreFiles` ourselves there.

Reply:
reworded the comment to be more clear about when the method is called and why we're setting things differently